tests/e2e/post_upgrade_test.go

```diff
@@ -0,0 +1,159 @@
+	vm1            string
+	vm2            string
+	vm1Name        string
+	vm2Name        string
+	volName1       string
+	volName2       string
+	containerName1 string
+	containerName2 string
+}
+
+func (s *PostUpgradeTestSuite) SetUpSuite(c *C) {
+	s.config = inputparams.GetTestConfig()
+	if s.config == nil {
+		c.Skip("Unable to retrieve test config, skipping basic tests")
+	}
+
+	s.esx = s.config.EsxHost
+	s.vm1 = s.config.DockerHosts[0]
+	s.vm2 = s.config.DockerHosts[1]
+	s.vm1Name = s.config.DockerHostNames[0]
+	s.vm2Name = s.config.DockerHostNames[1]
+}
+
+func (s *PostUpgradeTestSuite) SetUpTest(c *C) {
+	s.volName1 = "preUpgradeTestVol"
+	s.volName2 = "postUpgradeTestVol"
+	s.containerName1 = inputparams.GetUniqueContainerName(c.TestName())
+	s.containerName2 = inputparams.GetUniqueContainerName(c.TestName())
+}
+
+var _ = Suite(&PostUpgradeTestSuite{})
+
+// Test steps:
+// 1. Verify volume created in pre-upgrade test is available
+// 2. Verify the volume is detached
+// 3. Attach the volume from a new conatainer
+// 4. Verify volume status is attached
+// 5. Read data from that volume to make sure data written in pre-upgrade test remains same
+// 5. Write more data to that volume
+// 6. Read the data from the volume to make sure it matches the data written in previous step
+// 7. Stop the container and verify volume status is detached
+// 8. Remove the container
+// 9. Remove the volume created in pre-upgrade test
+// 10. Create a new volume
+// 11. Verify the status is detached
+// 12. Attach the new volume from a new container
+// 13. Verify the volume status is attached
+// 14. Write some data to the new volume
+// 15. Read the data from the new volume to make it matches the data written in previous step
+// 16. Stop the container
+// 17. Remove the container
+// 18. Remove the new volume
+func (s *PostUpgradeTestSuite) TestVolumeLifecycle(c *C) {
+	const (
+		testData  = "Hello World!"
+		testFile  = "hello.txt"
+		testData1 = "Hello World Again!"
+		testFile1 = "hello1.txt"
+		volPath   = "/vol"
+	)
+	misc.LogTestStart(c.TestName())
+
+	accessible := verification.CheckVolumeAvailability(s.vm1, s.volName1)
+	c.Assert(accessible, Equals, true, Commentf("Volume %s is not available on docker host %s",
+		s.volName1, s.vm1))
+
+	status := verification.VerifyDetachedStatus(s.volName1, s.vm1, s.esx)
+	c.Assert(status, Equals, true, Commentf("Volume %s is not detached", s.volName1))
+
+	out, err := dockercli.AttachVolume(s.vm1, s.volName1, s.containerName1)
+	c.Assert(err, IsNil, Commentf(out))
+
+	status = verification.VerifyAttachedStatus(s.volName1, s.vm1, s.esx)
+	c.Assert(status, Equals, true, Commentf("Volume %s is not attached", s.volName1))
+
+	out, err = dockercli.ReadFromContainer(s.vm1, s.containerName1, volPath, testFile)
+	c.Assert(err, IsNil, Commentf(out))
+	c.Assert(out, Equals, testData)
+
+	out, err = dockercli.WriteToContainer(s.vm1, s.containerName1, volPath, testFile1, testData1)
+	c.Assert(err, IsNil, Commentf(out))
+
```

Comment:
The correct test would be to "attach volume, write to container, stop container, attach volume again and read volume". This ensures that the container is getting back the same volume across different runs of the container. It verifies that the plugin is attaching the correct volume when requested.

An extension would be to write the volume on one VM, stop the container, run a container with the same volume on another VM and read back the data. That way we will be testing that the correct volume is attached each time, the volume is detached between the time one container writes the data on one VM and another reads the data on another VM (otherwise this won't work).

The write() and read() above will be cached in the guest OS memory and there is nothing thats testing the volume or whether the plugin and ESX service are working properly.

The whole test should be such that the actions of the plugin are tested.